db/memdb/db.go

```diff
@@ -0,0 +1,272 @@
+package memdb
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	dbm "github.com/cosmos/cosmos-sdk/db"
+	"github.com/google/btree"
+)
+
+const (
+	// The approximate number of items and children per B-tree node. Tuned with benchmarks.
+	bTreeDegree = 32
+)
+
+// MemDB is an in-memory database backend using a B-tree for storage.
+//
+// For performance reasons, all given and returned keys and values are pointers to the in-memory
+// database, so modifying them will cause the stored values to be modified as well. All DB methods
+// already specify that keys and values should be considered read-only, but this is especially
+// important with MemDB.
+//
+// Versioning is implemented by maintaining references to copy-on-write clones of the backing btree.
+//
+// TODO: Currently transactions do not detect write conflicts, so writers cannot be used concurrently.
+type MemDB struct {
+	btree       *btree.BTree            // Main contents
+	mtx         sync.RWMutex            // Guards version history
+	saved       map[uint64]*btree.BTree // Past versions
+	vmgr        *dbm.VersionManager     // Mirrors version keys
+	openWriters int32                   // Open writers
+}
+
+type dbTxn struct {
+	btree *btree.BTree
+	db    *MemDB
+}
+type dbWriter struct{ dbTxn }
+
+var _ dbm.DBConnection = (*MemDB)(nil)
+var _ dbm.DBReader = (*dbTxn)(nil)
+var _ dbm.DBWriter = (*dbWriter)(nil)
+var _ dbm.DBReadWriter = (*dbWriter)(nil)
+
+// item is a btree.Item with byte slices as keys and values
+type item struct {
+	key   []byte
+	value []byte
+}
+
+// NewDB creates a new in-memory database.
+func NewDB() *MemDB {
+	return &MemDB{
+		btree: btree.New(bTreeDegree),
+		saved: make(map[uint64]*btree.BTree),
+		vmgr:  dbm.NewVersionManager(nil),
+	}
+}
+
+func (db *MemDB) newTxn(tree *btree.BTree) dbTxn {
+	return dbTxn{tree, db}
+}
+
+// Close implements DB.
+// Close is a noop since for an in-memory database, we don't have a destination to flush
+// contents to nor do we want any data loss on invoking Close().
+// See the discussion in https://github.com/tendermint/tendermint/libs/pull/56
+func (db *MemDB) Close() error {
+	return nil
+}
+
+// Versions implements DBConnection.
+func (db *MemDB) Versions() (dbm.VersionSet, error) {
+	db.mtx.RLock()
+	defer db.mtx.RUnlock()
+	return db.vmgr, nil
```

Comment:
do we not want to return a copy of the map?

Reply:
It seems we already have a `Copy` method too.

Reply:
I opted to make this copy-on-write. `Versions` is called whenever version history is queried, and if the history becomes large, it could be expensive. So a `vmgr` is just copied over whenever versions are saved/deleted, and the caller can safely keep the old reference.

Reply:
`vmgr` is a pointer, so we don't make any object copy here